semantic-core/src/Analysis/ScopeGraph.hs

```diff
@@ -1,15 +1,130 @@
+import           Control.Effect.Fresh
+import           Control.Effect.Reader
+import           Control.Effect.State
+import           Control.Monad ((>=>))
+import           Data.File
+import           Data.Foldable (fold)
+import           Data.Function (fix)
+import           Data.List.NonEmpty
 import           Data.Loc
 import qualified Data.Map as Map
+import           Data.Name
+import           Data.Proxy
 import qualified Data.Set as Set
+import           Data.Text (Text)
+import           Data.Traversable (for)
+import           Prelude hiding (fail)
 
-data Entry = Entry
-  { entrySymbol :: String -- FIXME: Text
-  , entryLoc    :: Loc
+data Decl = Decl
+  { declSymbol :: Text
+  , declLoc    :: Loc
   }
+  deriving (Eq, Ord, Show)
 
-type ScopeGraph = Map.Map Entry (Set.Set Entry)
+newtype Ref = Ref Loc
+  deriving (Eq, Ord, Show)
+
+newtype ScopeGraph = ScopeGraph { unScopeGraph :: Map.Map Decl (Set.Set Ref) }
+  deriving (Eq, Ord, Show)
+
+instance Semigroup ScopeGraph where
+  ScopeGraph a <> ScopeGraph b = ScopeGraph (Map.unionWith (<>) a b)
+
+instance Monoid ScopeGraph where
+  mempty = ScopeGraph Map.empty
+
+scopeGraph
+  :: Ord term
+  => (forall sig m
+     .  (Carrier sig m, Member (Reader Loc) sig, MonadFail m)
+     => Analysis term User ScopeGraph m
+     -> (term -> m ScopeGraph)
+     -> (term -> m ScopeGraph)
+     )
+  -> [File term]
+  -> (Heap User ScopeGraph, [File (Either (Loc, String) ScopeGraph)])
+scopeGraph eval
+  = run
+  . runFresh
+  . runHeap
+  . traverse (runFile eval)
+
+runFile
+  :: ( Carrier sig m
+     , Effect sig
+     , Member Fresh sig
+     , Member (State (Heap User ScopeGraph)) sig
+     , Ord term
+     )
+  => (forall sig m
+     .  (Carrier sig m, Member (Reader Loc) sig, MonadFail m)
+     => Analysis term User ScopeGraph m
+     -> (term -> m ScopeGraph)
+     -> (term -> m ScopeGraph)
+     )
+  -> File term
+  -> m (File (Either (Loc, String) ScopeGraph))
+runFile eval file = traverse run file
+  where run = runReader (fileLoc file)
+            . runReader (Map.empty @User @Loc)
+            . runFailWithLoc
+            . fmap fold
+            . convergeTerm (Proxy @User) (fix (cacheTerm . eval scopeGraphAnalysis))
+
+-- FIXME: decompose into a product domain and two atomic domains
+scopeGraphAnalysis
+  :: ( Alternative m
+     , Carrier sig m
+     , Member (Reader Loc) sig
+     , Member (Reader (Map.Map User Loc)) sig
+     , Member (State (Heap User ScopeGraph)) sig
+     )
+  => Analysis term User ScopeGraph m
+scopeGraphAnalysis = Analysis{..}
+  where alloc = pure
+        bind name _ m = do
+          loc <- ask @Loc
+          local (Map.insert name loc) m
+        lookupEnv = pure . Just
+        deref addr = do
+          ref <- asks Ref
+          bindLoc <- asks (Map.lookup addr)
+          cell <- gets (Map.lookup addr >=> nonEmpty . Set.toList)
+          maybe (pure Nothing) (foldMapA (pure . Just . mappend (extendBinding addr ref bindLoc))) cell
```

Comment:
Because we’re modelling scope graphs as an abstract domain, we’re limited to producing them in the few methods which take or return a `value`, notably `deref`, `assign`, and `record`. In particular, we might prefer to build them in `lookupEnv` or `...` in order to preserve nested structure in the latter, but that would require out-of-band scope graph construction e.g. via a `Writer` effect.

Reply:
Can we pull out `(foldMapA (pure . Just . mappend (extendBinding addr ref bindLoc)))` into a let-bound function with a name? It’s a bit intimidating as is.

Reply:
Fixed in 0515d7c11.